raiden/waiting.py

```diff
@@ -42,6 +54,243 @@
+@dataclass
+class ChannelCoopSettleSuccess(ChannelStateCondition):
+    coop_settle_initiator_address: Address
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        if channel_state is None:
+            return False
+        if self.coop_settle_initiator_address is channel_state.our_state.address:
+            channel_end_state = channel_state.our_state
+        elif self.coop_settle_initiator_address is channel_state.partner_state.address:
+            channel_end_state = channel_state.partner_state
+        else:
+            raise ValueError("target_address must be one of the channel participants")
+        if (
+            channel_end_state.initiated_coop_settle is not None
+            and channel_end_state.initiated_coop_settle.transaction is not None
+        ):
+            return (
+                channel_end_state.initiated_coop_settle.transaction.result
+                is TransactionExecutionStatus.SUCCESS
+            )
+        return False
+
+
+@dataclass
+class And(ChannelStateCondition):
+    conditions: Iterable[ChannelStateCondition]
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        return all(condition(chain_state, channel_state) for condition in self.conditions)
+
+
+@dataclass
+class Or(ChannelStateCondition):
+    conditions: Iterable[ChannelStateCondition]
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        return any(condition(chain_state, channel_state) for condition in self.conditions)
+
+
+@dataclass
+class ChannelStateWaiter:
+    raiden: "RaidenService"
+    retry_timeout: float
+    token_network_registry_address: TokenNetworkRegistryAddress
+    token_address: TokenAddress
+    partner_address: Address
+
+    def _get_current_chain_state(self) -> ChainState:
+        return views.state_from_raiden(self.raiden)
+
+    def _get_channel_state(self, chain_state: ChainState) -> Optional[NettingChannelState]:
+        return _get_channel_state_by_partner_address(
+            chain_state,
+            self.token_network_registry_address,
+            self.token_address,
+            self.partner_address,
+        )
+
+    def _log_details(self, condition: ChannelStateCondition) -> Mapping[str, Any]:
+        log_details = {
+            "token_network_registry_address": self.token_network_registry_address,
+            "condition": str(condition),
+        }
+        return log_details
+
+    def wait(self, condition: ChannelStateCondition) -> None:
+        chain_state = self._get_current_chain_state()
+        while not condition(chain_state, self._get_channel_state(chain_state)):
+            assert self.raiden.is_running(), ALARM_TASK_ERROR_MSG
+            assert self.raiden.alarm.is_running(), ALARM_TASK_ERROR_MSG
+            log.debug(str(self), **self._log_details(condition))
```

Comment:
Readable yes, concise is debatable. This is how it looks like:

```ChannelStateWaiter(raiden=<RaidenService node:0x0417375797565580E623EB7E44D2061D1d7bf245>, retry_timeout=0.5, token_network_registry_address=b"\x9f\x1a\x0721La4\x1et\x12R\xc7'\n\x17\x11\xfaN\xdd", token_address=b'#jo\xc5hzeZv\x1f\xf4\xc9j\xdb\xb3\xcc\xf7\x13\xcbR', partner_address=b'\x88\xb0\x1b\xa8b*\xb9l\x86\x19w\x86\x11=\xff\x91\x9d\xc9v3')```

Reply:
Hm, I'd say that's borderline unreadable ;)
Esp. since it's all squished into the `event` field (when looking at it from the json log output perspective) and not split out into the key value portion.

Also does it show what it's waiting for? IMO that's the most important information. 
For example think about debugging an SP run. There filtering logs for `wait_for_deposit` is quite easy.

Reply:
I've changed the output to be more readable:
```
2021-08-19 08:26:02.417207 [debug    ] Waiting on channel             [raiden.waiting] condition=ChannelExists() node=0x0417375797565580E623EB7E44D2061D1d7bf245 partner_address=0x88b01bA8622aB96c86197786113dff919dc97633
2021-08-19 08:26:42.685264 [debug    ] Waiting on channel             [raiden.waiting] condition=ChannelInTargetStates(target_states=(<ChannelState.STATE_SETTLED: 'settled'>,)) node=0x0417375797565580E623EB7E44D2061D1d7bf245 partner_address=0x88b01bA8622aB96c86197786113dff919dc97633
```